From 63758efae8d4455235ecf6cf343a78a920e90257 Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Wed, 26 Oct 2016 15:05:34 -0400
Subject: [PATCH xserver] Revert "os/xdmcp: Just send XDMCP keepalive packets
 once every three minutes"

This reverts commit db1089eafc1c5371fa0030202de588d2e2b4f8e5.
---
 os/xdmcp.c | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/os/xdmcp.c b/os/xdmcp.c
index 5bdcbe9..6f9a7b2 100644
--- a/os/xdmcp.c
+++ b/os/xdmcp.c
@@ -84,6 +84,8 @@ static int req_socklen;
 static CARD32 SessionID;
 static CARD32 timeOutTime;
 static int timeOutRtx;
+static CARD32 defaultKeepaliveDormancy = XDM_DEF_DORMANCY;
+static CARD32 keepaliveDormancy = XDM_DEF_DORMANCY;
 static CARD16 DisplayNumber;
 static xdmcp_states XDM_INIT_STATE = XDM_OFF;
 
@@ -630,7 +632,6 @@ XdmcpOpenDisplay(int sock)
     if (state != XDM_AWAIT_MANAGE_RESPONSE)
         return;
     state = XDM_RUN_SESSION;
-    timeOutTime = GetTimeInMillis() + XDM_DEF_DORMANCY * 1000;
     sessionSocket = sock;
 }
 
@@ -688,6 +689,7 @@ XdmcpWakeupHandler(void *data,        /* unused */
                    int i, void *pReadmask)
 {
     fd_set *last_select_mask = (fd_set *) pReadmask;
+    fd_set devicesReadable;
 
     if (state == XDM_OFF)
         return;
@@ -702,6 +704,13 @@ XdmcpWakeupHandler(void *data,        /* unused */
             FD_CLR(xdmcpSocket6, last_select_mask);
         }
 #endif
+        XFD_ANDSET(&devicesReadable, last_select_mask, &EnabledDevices);
+        if (XFD_ANYSET(&devicesReadable)) {
+            if (state == XDM_RUN_SESSION)
+                keepaliveDormancy = defaultKeepaliveDormancy;
+        }
+        if (XFD_ANYSET(&AllClients) && state == XDM_RUN_SESSION)
+            timeOutTime = GetTimeInMillis() + keepaliveDormancy * 1000;
     }
     else if (timeOutTime && (int) (GetTimeInMillis() - timeOutTime) >= 0) {
         if (state == XDM_RUN_SESSION) {
@@ -1390,8 +1399,15 @@ recv_alive_msg(unsigned length)
     if (XdmcpReadCARD8(&buffer, &SessionRunning) &&
         XdmcpReadCARD32(&buffer, &AliveSessionID)) {
         if (SessionRunning && AliveSessionID == SessionID) {
+            /* backoff dormancy period */
             state = XDM_RUN_SESSION;
-            timeOutTime = GetTimeInMillis() + XDM_DEF_DORMANCY * 1000;
+            if ((GetTimeInMillis() - LastEventTime(XIAllDevices).milliseconds) >
+                keepaliveDormancy * 1000) {
+                keepaliveDormancy <<= 1;
+                if (keepaliveDormancy > XDM_MAX_DORMANCY)
+                    keepaliveDormancy = XDM_MAX_DORMANCY;
+            }
+            timeOutTime = GetTimeInMillis() + keepaliveDormancy * 1000;
         }
         else {
             XdmcpDeadSession("Alive response indicates session dead");
-- 
2.9.3

